secure-api-gateway-ob-uk-rs-server/src/main/java/com/forgerock/sapi/gateway/ob/uk/rs/server/api/obie/payment/v4_0_0/domesticscheduledpayments/DomesticScheduledPaymentsApiController.java

```diff
@@ -0,0 +1,280 @@
+    private final OBValidationService<OBWriteDomesticScheduled2ValidationContext> paymentValidator;
+    private final RefundAccountService refundAccountService;
+    private final IdempotentPaymentService<FRDomesticScheduledPaymentSubmission, FRWriteDomesticScheduled> idempotentPaymentService;
+
+    public DomesticScheduledPaymentsApiController(
+            DomesticScheduledPaymentSubmissionRepository paymentSubmissionRepository,
+            PaymentSubmissionValidator paymentSubmissionValidator,
+            @Qualifier("v4.0.0RestDomesticScheduledPaymentConsentStoreClient") DomesticScheduledPaymentConsentStoreClient consentStoreClient,
+            RefundAccountService refundAccountService,
+            OBValidationService<OBWriteDomesticScheduled2ValidationContext> paymentValidator
+    ) {
+        this.scheduledPaymentSubmissionRepository = paymentSubmissionRepository;
+        this.paymentSubmissionValidator = paymentSubmissionValidator;
+        this.consentStoreClient = consentStoreClient;
+        this.refundAccountService = refundAccountService;
+        this.paymentValidator = paymentValidator;
+        this.idempotentPaymentService = new SinglePaymentForConsentIdempotentPaymentService<>(paymentSubmissionRepository);
+    }
+
+    @Override
+    public ResponseEntity<OBWriteDomesticScheduledResponse5> createDomesticScheduledPayments(
+            String authorization,
+            String xIdempotencyKey,
+            String xJwsSignature,
+            OBWriteDomesticScheduled2 obWriteDomesticScheduled2,
+            String xFapiAuthDate,
+            String xFapiCustomerIpAddress,
+            String xFapiInteractionId,
+            String xCustomerUserAgent,
+            String apiClientId,
+            HttpServletRequest request,
+            Principal principal) throws OBErrorResponseException, OBErrorException {
+
+        logger.debug("Received payment submission: '{}'", obWriteDomesticScheduled2);
+
+        paymentSubmissionValidator.validateIdempotencyKey(xIdempotencyKey);
+
+        String consentId = obWriteDomesticScheduled2.getData().getConsentId();
+        logger.debug("Attempting to get consent: {}, clientId: {}", consentId, apiClientId);
+        final DomesticScheduledPaymentConsent consent = consentStoreClient.getConsent(consentId, apiClientId);
+        logger.debug("Got consent from store: {}", consent);
+
+        FRWriteDomesticScheduled frScheduledPayment = toFRWriteDomesticScheduled(obWriteDomesticScheduled2);
+        logger.trace("Converted to: '{}'", frScheduledPayment);
+
+        final Optional<FRDomesticScheduledPaymentSubmission> existingPayment =
+                idempotentPaymentService.findExistingPayment(frScheduledPayment, consentId, apiClientId, xIdempotencyKey);
+        if (existingPayment.isPresent()) {
+            logger.info("Payment submission is a replay of a previous payment, returning previously created payment for x-idempotencyKey: {}, consentId: {}",
```

Comment:
Does it need to be info? Maybe.

Is it ok to log `idempotencyKey` - or is that secret material? If so, maybe elsewhere too?

Reply:
Please see https://github.com/SecureApiGateway/secure-api-gateway-ob-uk-rs/pull/362#discussion_r1847940482.